### DISCUSSION TITLE
Caching date-based SQL queries

### DISCUSSION THREAD

**Discussion by pelguetat** [zilliztech/GPTCache#484] [Q&A]
So I've managed to get the cache working with Sqlite and Milvus against my Postgres db, and it works quite well. The issue I'm having is that for date based queries like: 
SELECT EXTRACT(MONTH FROM transaction_date) AS month, SUM(sale_amount) AS total_sales
FROM transactions
WHERE EXTRACT(YEAR FROM transaction_date) = 2023
GROUP BY month
ORDER BY total_sales DESC
LIMIT 5;

The cached will be hit even if the dates don't match, which obviously makes the query useless, but also you wouldn't want to have a cached query per possible date.
The solutions I'm thinking are:

1. Pass the cached response through an LLM to fix the date, which sorts of defeats the purpose of caching.
2. Passed the cached response through a function that replaces the dates. This could work but would get complex for multiple different query formats.
Any thoughts?

**Comment by SimFG** [comment]
Maybe you can get the date entered by the user in `similarity evaluation`, compare it with the date in the cache, and then if it is same, make other cache similar judgments, such as distance, etc.

**Comment by pelguetat** [comment]
The thing is, I want the cache to also work if the date is not exactly the
same, but the core of the query is the same.
So, if I ask, what was the revenue on march 2022, and then what was the
revenue in april 2023, it would still pull the same cached query, but
replace for the appropriate dates.

**Comment by SimFG** [comment]
@pelguetat got it. This feeling is a bit difficult to realize, because there is no way to figure out which part needs to be replaced through pure code, and it is also impossible to extract the value of the replacement. The ideal would probably require reasoning through a small llm.

**Comment by pelguetat** [comment]
I know, although it kind of defeats the purpose of caching if you use an
llm lol. I'm currently using langchain's sqlchain, so I'm thinking maybe I
just decompose that and intercept the query before it's performed, and use
spacy or something like that to rewrite it and then execute and continue
the flow.

**Comment by SimFG** [comment]
yes, you are right. This will be very troublesome if combined
